### ISSUE TITLE
btc transaction fees

### ISSUE THREAD

**Issue by nomisma-qt** [Coinomi/coinomi-android#154]
With latest versions 1.6.9 and 1.6.9.1 the btc transaction fees are not honored, and transactions are sent with random, low fees like 0.00013620 or 0.00011580 or whatever. This makes the app more or less useless, i now have several transactions that might never confirm.

There is no documentation on the dynamic fees, and the drop down selection box makes no sense as it says 'exchange fees' and the choices really mean nothing.

What is this silliness all about? And no reply from the support email.

**Comment by usmarine2141** [comment]
yes i agree!!!!! i sent bitcoin over 4 hours ago, and havent gotten it still!!!

**Comment by usmarine2141** [comment]
where are you getting your value for the coins at? right now on poloniex, btc is trading at 1400+ USD and zcash over 93 USD..

**Comment by nomisma-qt** [comment]
> where are you getting your value for the coins at? right now on poloniex, btc is trading at 1400+ USD and zcash over 93 USD..

I don't understand, i was talking about transaction fees for btc. You can calculate them at https://bitcoinfees.21.co/

I need to try to find an older version of coinomi to be able to use it, the worst thing is that their support emails goes unnoticed, people should be warned about the latest versions.

**Comment by usmarine2141** [comment]
yeah. so i wonder what happens to our BTC... does it get lost in the nether? or what

**Comment by nomisma-qt** [comment]
you can check the transaction fees on blockchain explorer, and compare them to https://bitcoinfees.21.co/

For example, one transaction was sent with  a fee of 60.256 sat/B which is way too low. i currently use a fee of 0.0006 btc per kilobyte

**Comment by usmarine2141** [comment]
yeah that is the default in coinomi .0006... and im still waiting over 3 hours

**Comment by usmarine2141** [comment]
https://poloniex.com/exchange#btc_zec

**Comment by nomisma-qt** [comment]
The transactions will probably confirm at some point, but it might be days or weeks. There are various options and services you can use to try to speed up the process, such as blockchain's pushtx feature:

http://themerkle.com/how-to-rebroadcast-an-electrum-transaction-using-blockchains-pushtx/

get the HEX format by adding ?format=hex after the url

also:

https://www.viabtc.com/tools/txaccelerator/

**Comment by nomisma-qt** [comment]
> yeah that is the default in coinomi .0006... and im still waiting over 3 hours

you need to check what fee was actually used by viewing the transaction on blockchain explorer, just press the transaction in coinomi and you see the option to view it on blockchain explorer

**Comment by usmarine2141** [comment]
https://blockchain.info/tx/aaeb76f760a8fbcc100189d2f560be26d80a11221caca27427301c20f771744b


Says so

**Comment by nomisma-qt** [comment]
no that says 60 sats per byte which translates to roughly 0.0001 btc/Kb

in bitcoinfees change the display setting to 'satoshis per byte' in the left upper corner

**Comment by nomisma-qt** [comment]
this is getting really silly, if you are inclined, you could try contacting the authors on facebook or linkedin or something, since they don't even bother reading their support email. To me the dev team seems incompetent which is a shame, as they have a potentially great product.

**Comment by usmarine2141** [comment]
I agree. Not sure I understand the pushtx feature. I did that viabit... does that even work?

**Comment by nomisma-qt** [comment]
modify the url like so:

https://blockchain.info/tx/aaeb76f760a8fbcc100189d2f560be26d80a11221caca27427301c20f771744b?format=hex

then copypaste the result into https://blockchain.info/pushtx

you should see 'transaction already exists' which means success.

**Comment by nomisma-qt** [comment]
with your fee of 60 bytes be prepared to wait for days or weeks atleast. it will not confirm anytime soon.

**Comment by usmarine2141** [comment]
Shit. Nothing I can do to speed it up at all?

**Comment by nomisma-qt** [comment]
usmarine2141 i don't use facebook personally, but if you could try contacting the dev team there and ask them to read their email and reply to this thread that would be nice.

**Comment by nomisma-qt** [comment]
> Shit. Nothing I can do to speed it up at all?

nope.

**Comment by usmarine2141** [comment]
I never had a problem past 3 days ago sending until today

**Comment by usmarine2141** [comment]
I don't use it either. The devs have responded here once to me.

**Comment by nomisma-qt** [comment]
yes as i said, this is a bug in the latest versions. i think 1.6.8 and older work fine.

**Comment by erasmospunk** [comment]
@nomisma-qt @usmarine2141 you can use the fees from the Settings if you select "Default" when sending. However if you should see higher fees, about 0.0005BTC for a smallish transaction.

When you select "Normal", "Priority", "Low Priority" when sending, does the fee change?

**Comment by nomisma-qt** [comment]
i sent with both default and priority, it made no difference.

**Comment by nomisma-qt** [comment]
i suggest you actually try it yourself, since it's not very funny to test it only to discover all transactions are sent with some arbitrary low fee and then being unconfirmed for days.

You think it would be a good idea to actually test your code before releasing it yourself?

**Comment by nomisma-qt** [comment]
Please be aware that google play has version 1.6.8, which i believe works, this is the downloadable APK from the website. versions 1.6.9 and 1.6.9.1

**Comment by erasmospunk** [comment]
@nomisma-qt you don't need to actually send anything, just check the confirmation screen after pressing "Send", without pressing "Confirm". There is a drop-down menu next to "Change fees" and by default "Normal" is selected, now that I am trying it I'm getting 0.45USD for a small transaction, 0.55USD for "Priority" and for "Default" (that is set to 60sat/byte) I get 0.19USD.

I cannot reproduce the issue you are having. What is the version of your wallet?

Also keep in mind that the fees are calculated per byte, not per kilobyte like in the past. For example the above transaction is about 230 bytes and the "Normal" fee is about 0.00033BTC so about 143sat/byte that should be enough to confirm in about 2 hours and "Priority" is about 176sat/byte that should confirm in a couple of blocks.

**Comment by nomisma-qt** [comment]
my version is 1.6.9.1

First of all, 60sat/byte is way too low. You need to check https://bitcoinfees.21.co/

where you can see that at this moment 141 sat is the minimum if you want to do a normal transaction.

There might be days where this might be much higher, which is why 'normal', 'default' and 'priority' don't mean anything. A few weeks ago there was a situation where your 'priority' would have taken days to confirm.

I don't have btc to test what you described, but i'm pretty sure there is no way to see the actual fee used, however expressed, before actually sending. the most common transaction size is 226 bytes. Thus from the site above:

"The fastest and cheapest transaction fee is currently 160 satoshis/byte, shown in green at the top.
For the median transaction size of 226 bytes, this results in a fee of 36,160 satoshis."

Anyways, the issue is, i have a fee of 0.0006 in settings, and despite that my transactions were sent with weird arbitrary fees like 0.00026150 and 0.00011580 BTC, which make no sense at all, even when using 'default'

**Comment by nomisma-qt** [comment]
but if you could test it, set the fee to 0.0006 in settings, then send a small amount using both 'default' and 'priority', and then post the blockchain transaction urls here. Then we can further figure this out..

**Comment by nomisma-qt** [comment]
do the above with versions 1.6.9 and 1.6.9.1

**Comment by nomisma-qt** [comment]
to make my point, if you have 60 sats as 'normal' then you will soon have thousands of angry users with transactions that take days to confirm, which will basically make a mockery of the coinomi wallet.

**Comment by nomisma-qt** [comment]
on the above site, if you want the correct transaction fee in the format used in settings, you need to change the display to 'BTC per byte' in the upper left corner, and then multiply that with the transaction size, thus at the moment it would be 0.0000014 * 226 = 0.0003164

But ofcourse, for larger transaction the size will be much bigger than 226, resulting in a situation where a large transaction would still take days to confirm. Not a pleasant situation for anyone to be wondering if their 50 btc transaction will ever confirm. There is no way to tell what fee will actually be used in coinomi before actually pressing send, and by then it will be too late to change.

**Comment by nomisma-qt** [comment]
further commenting, if it would be possible to have coinomi calculate the actual transaction size before hitting send, it would be an awesome feature, because again, a few weeks ago there was a situation where 0.0006 was the minimum for a normal transaction with 10 minute block times, reddit was filled with thousands of desperate messages from people with unconfirmed transactions etc etc..

At the moment there are very few wallets that actually have this feature, and it is highly likely the situation will repeat at some point, so in that sense having such a feature could rocket coinomi into one of the most popular android wallets of all times.

**Comment by nomisma-qt** [comment]
and still to clarify, in versions 1.6.8 and earlier, the fee in settings was the fee actually used for the transaction, that way i could always check bitcoinfees21 before sending to ensure the fee was high enough.

But as i said at the beginning of this thread, with version 1.6.9 and later, this setting was not honored even when sending with 'default' So if you could do a test send with 'default' and another one with 'priority' and then post the blockchain.info urls here it should confirm my original complaint..

**Comment by erasmospunk** [comment]
@nomisma-qt the fee that you see in the settings is unrelated to the dynamic fees ("Low priority", "Normal" and "Priority"). The dynamic fees take from 3 blocks ("Priority") to 24 blocks ("Low priority"). Here is a tx with "Normal" priority:

https://blockchain.info/tx/eecbb0d0384fe16e328149118711008445db4f9b9c07b0f61cb033f8ce5d02e6

The "Default" fee is the fees a users sets in the Settings:

https://blockchain.info/tx/d6aa069f709ee8481ebc5deaf9478eae2c56f2db6fe3d347bfb273487cd4a34f

In this transaction, I purposely set low fees and it took about a week to confirm.

The wallet already calculates the transaction size when in the confirmation screen and it shows the final fees.

So if you got a transaction with a fee of 0.00026150 and the size of the transaction was 225 bytes (1 input and 2 outputs) then the fee is about 116sat/byte. The dynamic fee, is more that enough for fast confirmations. I just sent a transaction using the "Default" fee of 60sat/byte and it confirmed on the next block (however it will not always confirm that fast)

https://blockchain.info/tx/e789add149abcfc9a12828dece0402dbbcf184c3fee3e12074734a304d079006

**Comment by nomisma-qt** [comment]
ok i have to test again after i have some btc to play with. But, my original complaint was really that with 'default', the fee in settings was actually not honored, instead the fee was something else. this was with 1.6.9.1 but i can't confirm. Are you sure that if you set the fee to 0.0006 in version 1.6.9.1, and then send using 'default' that it's the actual fee used?

**Comment by nomisma-qt** [comment]
also to be sure, my version 1.6.9.1 was downloaded from the site, and i checked the SHA256 hash to make sure it was legit. Have you checked to see that the site hasn't been hacked? Can you verify the SHA256 value on your site?

**Comment by nomisma-qt** [comment]
i'm sending some btc to my coinomi wallet now to confirm..

**Comment by erasmospunk** [comment]
@nomisma-qt the Bitcoin fee calculation algorithm changed in 1.6.8 to be per-bytes instead of per-kilobytes (1000 bytes). So if the default fee is 0.0006BTC per KB (or 60sat/byte) and your transaction is 227 bytes you get a fee of 0.0001362 BTC while in pre v1.6.8 you would pay a fee of 0.0006 BTC (the 227 byte transaction is rounded up to 1000 bytes).

This is the case for the [transaction](https://blockchain.info/tx/e789add149abcfc9a12828dece0402dbbcf184c3fee3e12074734a304d079006) that I just sent, with the "Default" fee of 60sat/byte I got a fee of 0.0001362 BTC. Notice that the actual transaction is 226 bytes instead of 227 because the signature can vary in size.

**Comment by nomisma-qt** [comment]
ok, but personally i never want to send anything with 0.0001 BTC/KB, because that will result in a transaction that takes days to confirm. This is also what will happen to most users, resulting in alot of crying. So i don't understand the logic, because people don't want to wait for days.

**Comment by nomisma-qt** [comment]
just testing now, in settings i have 0.0006. then i choose 'default' before sending, and it gives me 0.00013620. But i don't want that, i want 0.0006, because otherwise, i will once again wait for DAYS

what gives?

**Comment by nomisma-qt** [comment]
also if i then switch to 'priority' or 'normal' or 'low priority' before sending, the fee displayed does not change, it remains 0.00013620.

if i go back and forth between the first page with send button, and the second page with the dropdown, sometimes changing the 'change fees' does nothing, sometimes it goes back a page, sometimes fee is displayed, sometimes it's not

there's no logic here

and also, there is NO documentation about this, except for this very thread now.

**Comment by nomisma-qt** [comment]
also again, i am unable to force the fee to be 0.0006, even using priority, or default

i don't want to send anything with 0.0001, at all. never. But i cannot make sure i get 0.0006. I can't send btc using that fee, whatever i do.

**Comment by nomisma-qt** [comment]
if you set the fee to 0.0006 in settings, tell me, how to send a btc transaction using that xact fee? How do you do it? for me, using 'default' gives me 0.00013620.

But i don't want that, i want 0.0006!

**Comment by nomisma-qt** [comment]
i mean with this setup, i can never be sure that my transaction will confirm in 60 minutes, i will always be afraid that it will get stuck for days, just like happened to me and usmarine2141 resulting in this thread. 0.0001360 is not a fee any sane person would use right now. The last 3 transactions i made with 1.6.9.1 all took almost 3 days to confirm!

**Comment by erasmospunk** [comment]
@nomisma-qt if you change the fees to "Priority" and the Transaction fee remains the same, it is a bug. For some reason your wallet is not getting the fee recommendation from the server, maybe a slow connection? If you wait a little bit, does it change? We will investigate this.

Regarding the fee you can set in the settings, it isn't the final transaction fee, that one depends on the transaction size. The transaction size depends on factors that I will not explain here. So in your case the transaction size is 227 bytes so 0.0006 / 1000 * 227 = 0.0001362. If you want to set the fee in the settings to 0.0015 BTC per KB (or 150sat/byte) and you will get a transaction fee of 0.0003405 BTC.

**Comment by nomisma-qt** [comment]
oh right, now this is starting to make some sense.. The connection is slow because i have forced coinomi through tor only using iptables.

So this means i have to set the default transaction fee to something, that will then correspond to a final sum of 0.0006 taking into account the size. With the most common size of 226 bytes that is some number..

Still, it's a bit complicated if you're standing on the street after having 10 beers with the guys and need to pay for a cab quickly, my old brain would not be able to compute easily what i should do to make sure i won't be left standing in the rain.

Would it be possible to have a 'legacy fee' setting that would ensure 0.0006 or whatever the flavour of the month happens to be without being a math wizard?

This does clarify alot, i just wish you would have the time to document this better. Thanks for the interesting convo though..

**Comment by jonerys** [comment]
My problem is worst. I can't send BTC. Coinomi doesn't calculate the fee. I tried to reinstall 2x. Try in many connections. The error stills the same: "Error executing this action..." But if I put all my btc wallet funds the error doesn't happen. Very weird.

**Comment by nomisma-qt** [comment]
look Mr. Author, this isn't working.
Yif your client is dependent on dynamically calculated fees not stored locally, you must make sure all clients can access that data from anywhere in the world over any exit point like toor, cleartext etc

not a small task. The system used in 1.6.9.x is not going to work.

**Comment by nomisma-qt** [comment]
damnit i need to send some btc now and can't get the client to give any other fee than 0.00013620

NO GOOD

**Comment by usmarine2141** [comment]
Jonerys uninstall and reinstall BE SURE TO WRITE DOWN YOUR KEY

**Comment by nomisma-qt** [comment]
i need link to 1.6.8 apk

**Comment by nomisma-qt** [comment]
this is ridiculous

**Comment by usmarine2141** [comment]
Hate how they tell us to email support. Nobody answers. Only reason this is open is cause to many people have seen it.

**Comment by usmarine2141** [comment]
Yep... I'm about to uninstall

**Comment by nomisma-qt** [comment]
look Mr. Developer, coinomi version 1.6.9 is a disaster, it will absolutely not work.

**Comment by nomisma-qt** [comment]
on top of being sure every client has http access to the data, it also needs to work without DNS.

**Comment by nomisma-qt** [comment]
current situation is this: in order to send a btc transaction with a higher fee than 0.00011, is impossible with version 1.6.9.x, no matter the settings or drop down choice.

apk of version prior to 1.6.9 nowhere to be found

only recourse was to restore a wallet of version 1.6.5 on an old phone (which i had by luck), set the btc transaction fee in settings to 0.0006

and make a transaction with the fee of 261 sats or minimum required according to https://bitcoinfees.21.co/ at the moment of typing this.

**Comment by nomisma-qt** [comment]
not my job to investigate, but most likely cause is A: the host from which the dynamic fee numbers are pulled has banned tor exit nodes and B: pulling the numbers requires access to a DNS server.

Both A and B are total show stoppers, both separately or together, unless options provided to set them manually. Otherwise coinomi can never be used globally, only in regions where the internet is under complete government control and spy agency oversight.

**Comment by jonerys** [comment]
@usmarine2141 I already do this bro, no effect

**Comment by jonerys** [comment]
@nomisma-qt https://coinomistorage.blob.core.windows.net/binaries/coinomi-1.6.7.apk

1.6.8 doesn't work too, if u want to try 1.6.8 just change the version in URL

btw I moved my BTC wallet to mycelium, I can't wait they fix that. No support answers

**Comment by nomisma-qt** [comment]
Thanks for the link jonerys

**Comment by nomisma-qt** [comment]
the latest version 1.7.0.1 seems to work better, choosing default for transaction fee uses the value in the settings, and choosing priority calculates the fees pretty close to my manual settings which i calculated using bitcoinfees21.com

The only thing to remember is the new way to calculate fees, basically you need to pick your desired tx fee in satoshi/byte, and then divide that number with 100000, so as of writing this bitcoinfees21 says:

"The fastest and cheapest transaction fee is currently 360 satoshis/byte"

so the fee to ensure quick confirmations would be 360 / 100000 = 0.0036

the priority setting comes close to this, so good work devs!

I also understand the devs have had some issue with their email, which is being addressed. All in all it looks much better now, except maybe for documentation, because afaik this thread is the only documentation about the tx fee issues anywhere, a nice big link to explain this on coinomi's webpage would be a good idea.

Anyways thx to dev team for working hard on this, infact, i believe coinomi now has one of the more advances tx fee system out there, i don't know about mycelium though.
